File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        self.res = []


# 2,2,2,2
        def dfs(nums,target,lst):
            if target == 0:
                self.res.append(lst.copy())
                return
            if target < 0:
                return
            if len(nums) == 0:
                return
            
            lst.append(nums[0])
            dfs(nums,target-nums[0],lst)

            lst.pop()
            dfs(nums[1:],target,lst)
            return
        dfs(nums,target,[])
        return self.res

File: Data Structures & Algorithms/permutations/submission-1.cpp
#include <vector>
class Solution {
private:
std::vector<std::vector<int>> res;
std::vector<int> nums;

public:
    vector<vector<int>> permute(vector<int>& nums) {
        std::vector<int> curr_lst;
        this->nums = std::move(nums);
        dfs(curr_lst,0);
        return res;
    }

    void dfs(std::vector<int> curr_lst, int i){
        if (i == nums.size()){
            res.push_back(curr_lst);
            return;
        }

        for (int j=0;j<curr_lst.size()+1;j++){
            curr_lst.insert(curr_lst.begin()+j,nums[i]);
            dfs(curr_lst,i+1);
            curr_lst.erase(curr_lst.begin()+j);
        }

    }
};
